beacon-chain/p2p/peers/status.go

```diff
@@ -0,0 +1,311 @@
+//
+// For convenience, there are two aggregate states expressed in functions:
+//
+// - active if we are connecting or connected
+// - inactive if we are disconnecting or disconnected
+//
+// Peer information is persistent for the run of the service.  This allows for collection of useful long-term statistics such as
+// number of bad responses obtained from the peer, giving the basis for decisions to not talk to known-bad peers.
+package peers
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/libp2p/go-libp2p-core/network"
+	"github.com/libp2p/go-libp2p-core/peer"
+	ma "github.com/multiformats/go-multiaddr"
+	pb "github.com/prysmaticlabs/prysm/proto/beacon/p2p/v1"
+	"github.com/prysmaticlabs/prysm/shared/roughtime"
+)
+
+// PeerConnectionState is the state of the connection.
+type PeerConnectionState int
+
+const (
+	// PeerDisconnected means there is no connection to the peer.
+	PeerDisconnected PeerConnectionState = iota
+	// PeerConnecting means there is an on-going attempt to connect to the peer.
+	PeerConnecting
+	// PeerConnected means the peer has an active connection.
+	PeerConnected
+	// PeerDisconnecting means there is an on-going attempt to disconnect from the peer.
+	PeerDisconnecting
+)
+
+// Status is the structure holding the peer status information.
+type Status struct {
+	lock            sync.RWMutex
+	maxBadResponses int
+	status          map[peer.ID]*peerStatus
+}
+
+// peerStatus is the status of an individual peer at the protocol level.
+type peerStatus struct {
+	address               ma.Multiaddr
+	direction             network.Direction
+	peerState             PeerConnectionState
+	chainState            *pb.Status
+	chainStateLastUpdated time.Time
+	badResponses          int
+}
+
+// NewStatus creates a new status entity.
+func NewStatus(maxBadResponses int) *Status {
+	return &Status{
+		maxBadResponses: maxBadResponses,
+		status:          make(map[peer.ID]*peerStatus),
+	}
+}
+
+// MaxBadResponses returns the maximum number of bad responses a peer can provide before it is considered bad.
+func (p *Status) MaxBadResponses() int {
+	return p.maxBadResponses
+}
+
+// Add adds a peer.  This will error if the peer already exists.
+func (p *Status) Add(pid peer.ID, address ma.Multiaddr, direction network.Direction) error {
+	p.lock.Lock()
+	defer p.lock.Unlock()
+
+	if _, ok := p.status[pid]; ok {
+		return fmt.Errorf("peer %v already exists", pid)
+	}
+
+	status := &peerStatus{
+		address:   address,
+		direction: direction,
+		peerState: PeerDisconnected,
+	}
+
+	p.status[pid] = status
+	return nil
+}
+
+// Address returns the multiaddress of the given remote peer.
+// This will error if the peer does not exist.
+func (p *Status) Address(pid peer.ID) (ma.Multiaddr, error) {
+	p.lock.RLock()
+	defer p.lock.RUnlock()
+
+	if status, ok := p.status[pid]; ok {
+		return status.address, nil
+	}
+	return nil, fmt.Errorf("peer %v unknown", pid)
```

Comment:
Same feedback with all of these errors. Rather than using formatted errors, you can return global errors and check these certain cases. 

Also peer.ID is ugly unless you call `.Pretty()`

Reply:
I'll make them globals.

FYI peer.ID autoformats to `Pretty()`:
```
func (id ID) String() string {
      return id.Pretty()
}
```

Reply:
Now using global errors.

Reply:
Ah, does it call string with %v? It never got a nice message without calling Pretty. Thanks